Out-of-bounds atom indices in bond table

ptosco opened the pull request:
@rohdebe1
Currently out-of-bounds atom indices in bond table cause a segmentation fault due to unallocated memory being accessed on write.
To avoid the problem I have added a check in `struchk.c` for such out-of-bounds atom indices.
Specifically, I incurred into the segfault when parsing the following misaligned CTAB:
```

     RDKit          2D

  3  2  0  0  0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2990    0.7500    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981   -0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
 1  2  1  0
  2  3  1  0
M  END
```
Due to zero padding taking place in `reaccsio.c`, the atom indices in the first misaligned bond line were being parsed as `10` and `20`, respectively.
While the 1st line is misaligned, it can be read correctly if zero padding is not performed. Zero padding is actually unnecessary as `sscanf` will simply skip spaces.

Commit:
- added a check for out-of-bounds atom indices in the bond table
- removed zero padding as it is unnecessary and potentially harmful

rohdebe1 commented:
Unfortunately, the MOL-file format was defined with Fortran semantics, i.e. fixed field numbers and zero padding of blanks.
Just dropping the padding and relying on space separated numbers is not going to cut it. A better way to safeguard against non-aligned numbers might be to check before access when reading atom numbers, since the numbers of atoms and bonds should be known at that time. It might also be possible to spot misalignment in other ways, but violating the format definition isn't an option, I think.
Could you tell me how you came across a file that showed misalignment? Such a misalignment is usually just an indication of more dubious input to come from that source ;-)

Commit:
interpret spaces as Fortran does

ptosco commented:
> Unfortunately, the MOL-file format was defined with Fortran semantics, i.e. fixed field numbers

Hi Bernd, I am well aware of that.

> and zero padding of blanks.

I disagree. The way Avalon was padding the atom index fields is *not* what Fortran does. Fortran will not replace space with zero _after_ a digit, as that would mean turning number `x` into `x*10^n` where `n` is the number of trailing spaces replaced with 0.

See the following sample Fortran program `read_test.f`:
```fortran
      program main
      implicit none
      
      character(len=200), dimension(9) :: inputarr
      character outstr*1000
      integer i
      integer atom_idx1
      integer atom_idx2
      integer bond_type
      integer stereo_symbol
      integer dummy
      integer topography
      integer reaction_mark
  
      atom_idx1 = -1
      atom_idx2 = -1
      bond_type = -1
      stereo_symbol = -1
      dummy = -1
      topography = -1
      reaction_mark = -1
      inputarr(1) = ' 1  2  1  0'
      inputarr(2) = '  1  2  1  0'
      inputarr(3) = '100200  1  0  0'
      inputarr(4) = ' 10 20     1  2'
      inputarr(5) = '    20'
      inputarr(6) = '        1  0  0'
      inputarr(7) = '      1 2 3 4'
      inputarr(8) = ' 10'
      inputarr(9) = '   '
      do i = 1, 9
        write(*, '(3a)') 'bond_line: ''', trim(inputarr(i)), ''''
        read(inputarr(i), '(7i3)') atom_idx1, atom_idx2, bond_type,
     +    stereo_symbol, dummy, topography, reaction_mark
        write(outstr, '(8(a,i3))') 'i: ', i, ', atom_idx1: ', atom_idx1,
     +    ', atom_idx2: ', atom_idx2, ', bond_type: ', bond_type,
     +    ', stereo_symbol: ', stereo_symbol, ', dummy: ', dummy,
     +    ', topography: ', topography, ', reaction_mark: ',
     +    reaction_mark
        write(*, '(a,/)') trim(outstr)
      end do
      end
```
If you build the above and run it, you will see that zero padding does not take place as Avalon was doing it:
```
$ gfortran -Wall read_test.f -o read_test
$ ./read_test
bond_line: ' 1  2  1  0'
i:   1, atom_idx1:   1, atom_idx2:   2, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '  1  2  1  0'
i:   2, atom_idx1:   1, atom_idx2:   2, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '100200  1  0  0'
i:   3, atom_idx1: 100, atom_idx2: 200, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: ' 10 20     1  2'
i:   4, atom_idx1:  10, atom_idx2:  20, bond_type:   0, stereo_symbol:   1, dummy:   2, topography:   0, reaction_mark:   0

bond_line: '    20'
i:   5, atom_idx1:   0, atom_idx2:  20, bond_type:   0, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '        1  0  0'
i:   6, atom_idx1:   0, atom_idx2:   0, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '      1 2 3 4'
i:   7, atom_idx1:   0, atom_idx2:   0, bond_type:  12, stereo_symbol:   3, dummy:   4, topography:   0, reaction_mark:   0

bond_line: ' 10'
i:   8, atom_idx1:  10, atom_idx2:   0, bond_type:   0, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: ''
i:   9, atom_idx1:   0, atom_idx2:   0, bond_type:   0, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0
```
Example `i:    1` shows that `' 1 '` is parsed as such rather than being zero-padded to `'010'` as Avalon does, which yields `10`.
One thing I ignored is that `1 2` is parsed as `12`, i.e. spaces are just ignored, no matter if they are at the beginning, at the end or in between digits. I therefore updated my PR accordingly.

Now the C code I extracted from my PR yields exactly the same results as the Fortran code above:
```C
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define BUFSIZE 200
#define MAX_BONDLINE_FIELDS 7
#define BONDLINE_FIELD_LEN 3

int atom_idx1;
int atom_idx2;
int bond_type;
int stereo_symbol;
int dummy;
int topography;
int reaction_mark;

int parse_bond_line(const char *bond_line) {
    char buffer[BONDLINE_FIELD_LEN + 1], c;
    int *ptrarray[MAX_BONDLINE_FIELDS];
    int i, j, k, pos, n_chars;
    int nitems;
    int bond_line_len;

    atom_idx1 = 0;
    atom_idx2 = 0;
    bond_type = 0;
    stereo_symbol = 0;
    dummy = 0;
    topography = 0;
    reaction_mark = 0;
    ptrarray[0] = &atom_idx1;
    ptrarray[1] = &atom_idx2;
    ptrarray[2] = &bond_type;
    ptrarray[3] = &stereo_symbol;
    ptrarray[4] = &dummy;
    ptrarray[5] = &topography;
    ptrarray[6] = &reaction_mark;
    bond_line_len = strlen(bond_line);
    nitems = bond_line_len ? (bond_line_len - 1) / 3 + 1 : 0;
    if (nitems > MAX_BONDLINE_FIELDS) {
        nitems = MAX_BONDLINE_FIELDS;
    }
    for (i = 0; i < nitems; ++i) {
        pos = i * 3;
        memset(buffer, 0, BONDLINE_FIELD_LEN + 1);
        n_chars = bond_line_len - pos;
        if (n_chars > 3) {
            n_chars = 3;
        }
        for (j = 0, k = 0; j < n_chars; ++j) {
            c = bond_line[pos + j];
            if (c != ' ') {
                buffer[k++] = c;
            }
        }
        sscanf(buffer, "%3d", ptrarray[i]);
    }
    printf("bond_line: '%s'\n", bond_line);
    return nitems;
}

int main() {
    const char *bond_line[] = {
      " 1  2  1  0",
      "  1  2  1  0",
      "100200  1  0  0",
      " 10 20     1  2",
      "    20",
      "        1  0  0",
      "      1 2 3 4",
      " 10",
      "   ",
    };
    int i;
    for (i = 0; i < sizeof(bond_line) / sizeof(bond_line[0]); ++i) {
        parse_bond_line(bond_line[i]);
        printf("i: %3d, atom_idx1: %3d, atom_idx2: %3d, bond_type: %3d, stereo_symbol: %3d, dummy: %3d, topography: %3d, reaction_mark: %3d\n",
        i + 1, atom_idx1, atom_idx2, bond_type, stereo_symbol, dummy, topography, reaction_mark);
    }
    return 0;
}

$ gcc -Wall sscanf_test.c -o sscanf_test
$ ./sscanf_test
bond_line: ' 1  2  1  0'
i:   1, atom_idx1:   1, atom_idx2:   2, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '  1  2  1  0'
i:   2, atom_idx1:   1, atom_idx2:   2, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '100200  1  0  0'
i:   3, atom_idx1: 100, atom_idx2: 200, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: ' 10 20     1  2'
i:   4, atom_idx1:  10, atom_idx2:  20, bond_type:   0, stereo_symbol:   1, dummy:   2, topography:   0, reaction_mark:   0

bond_line: '    20'
i:   5, atom_idx1:   0, atom_idx2:  20, bond_type:   0, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '        1  0  0'
i:   6, atom_idx1:   0, atom_idx2:   0, bond_type:   1, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '      1 2 3 4'
i:   7, atom_idx1:   0, atom_idx2:   0, bond_type:  12, stereo_symbol:   3, dummy:   4, topography:   0, reaction_mark:   0

bond_line: ' 10'
i:   8, atom_idx1:  10, atom_idx2:   0, bond_type:   0, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0

bond_line: '   '
i:   9, atom_idx1:   0, atom_idx2:   0, bond_type:   0, stereo_symbol:   0, dummy:   0, topography:   0, reaction_mark:   0
```

> relying on space separated numbers is not going to cut it.

Again, I am well aware of this. That's not what my PR is about.

> A better way to safeguard against non-aligned numbers might be to check before access when reading atom numbers, since the numbers of atoms and bonds should be known at that time

Indeed. I have also added a check for that in `struchk.c`. In fact, at the moment attempting to parse this CTAB (which has no misalignment, just bogus atom indices in the bond table) with Avalon will cause a segfault, which should not happen:
```

     RDKit          2D

  3  2  0  0  0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2990    0.7500    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981   -0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
 10 20  1  0
  2  3  1  0
M  END
```
My PR avoids the segfault.

Commit:
replaced 3 with BONDLINE_FIELD_LEN
